Stepper: Redirect user to the login page using the correct language

gabrielcaires opened the pull request:
## Proposed Changes

* Redirect the user to sign the flow using the lang defined on the flow path (E.g `/setup/migration-signup/de` )
* Fix the redirect to `/setup/:flowName` to the first step to include the lang
* Quickly redirect the user to the `site-setup` flow when there is no flow on the URL.
* Centralize the logic to get the location on the `'ui.language.localeSlug'
* Improve logic to determine the flow step and locale based on the URL. 
* Update flows to use useFlowLocale instead of useLocale


> [!NOTE]  
> There is a possible race condition so you might see the steps rendering in English before rendering the selected language. It will be fixed as part of other PR.


## Why are these changes being made?
Nowadays when the user accesses/setup/:flowName/:lang we remove the lang part when we redirect the user to the first flow step. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Using a not logged-in browser, access `/setup/migration-signup/de`
* Check if the signup page is in German
* Use a random email on the email sign-up to create a new account
* Check if you see the steps in German.


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

gabrielcaires commented on client/package.json:
```diff
@@ -189,7 +189,7 @@
 		"react-modal": "^3.16.1",
 		"react-redux": "^8.1.3",
-		"react-router-dom": "^6.10.0",
+		"react-router-dom": "^6.23.1",
```
It is necessary due to this issue https://github.com/remix-run/react-router/pull/10768.
Version 6.10.0 doesn't support optional parameters (:lang?) on the useMatch hook.

daledupreez commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -170,7 +174,7 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
 					<Route
 						key={ step.slug }
-						path={ `/${ flow.variantSlug ?? flow.name }/${ step.slug }` }
+						path={ `/${ flow.variantSlug ?? flow.name }/${ step.slug }/:lang?` }
```
I have a concern: this feels like it will no longer match `/:flow/:step` without the trailing `/`, as I am not sure if the optionality applies to the `/`.

gabrielcaires replied:
> I am not sure if the optionality applies to the /.

Yes, it is applicable, the regular /:flow/:step is working properly.

gabrielcaires committed:
Improve redirect detection when there is no flow available

Site setup is the default flow when no flow is assigned, but it was loading the full flow and steps before redirect the user to the first flow step. It was causing edge cases because the flow name was not present on the url.

gabrielcaires committed:
Remove unused hook

I could do it in another PR but my changes were causing failing on hook tests, so I decided remove as part this PR.

gabrielcaires committed:
Improve redirection logic when the step name is not available

When the `/setup/flowName/es` is used, we were removing the locale param from URL, it was causing racing conditions because we could have the setupLocale code not capturing the locale code and missing this information.

gabrielcaires committed:
Use calypso store as our locale source of truth for the steps

The useLocale is the last one to be updated with the desired locale while the calypso store is the first one that have this information, so is prefereable to get the information from there.

gabrielcaires committed:
Move stepper navigation to a hook and use locale param properly

gabrielcaires committed:
Update more places to use useFlowLocale instead of useLocale

gabrielcaires committed:
Add support to access steps with lang directly

gabrielcaires commented on client/landing/stepper/declarative-flow/internals/hooks/use-flow-navigation/index.tsx:
```diff
@@ -0,0 +1,79 @@
+ * @returns
+ */
+
+export const useFlowNavigation = ( stepPaths: string[] ): FlowNavigation => {
```
@daledupreez  @omarjackman  Do you know the reason we need to store the stepPaths? 
I couldn't find references on the codebase.

alshakero replied:
The `stepPaths` can go. It seems to have started since [the very first commit to Stepper.](https://github.com/Automattic/wp-calypso/blob/b57297fac00a56cf103f3235e33f6636ddca1f49/client/landing/stepper/declarative-flow/internals/index.tsx#L22) And it most likely was just a mistake.

gabrielcaires replied:
Thanks! Removed here [aef0abf](https://github.com/Automattic/wp-calypso/pull/91935/commits/aef0abfe8e97a0de922621a079f1374562d3ce23)

alshakero commented on client/landing/stepper/declarative-flow/internals/hooks/use-flow-navigation/index.tsx:
```diff
@@ -0,0 +1,79 @@
+
+		const queryParams = ! hasQueryParams ? currentSearchParams : null;
+
+		setStepData( {
+			path: nextStep,
+			intent: intent,
+			previousStep: currentStep,
+			...extraData,
+		} );
```
Not related: I really want to get rid of this. It's code smell. Luckily, it's used a few times only and most of the uses don't really need it.

gabrielcaires replied:
Do you mean the full setStepData call or just the intent?

alshakero replied:
Step data in general can probably be removed but needs some care.

gabrielcaires replied:
Yes, let's leave it for future PRs

alshakero commented on client/landing/stepper/declarative-flow/internals/hooks/use-flow-navigation/index.tsx:
```diff
@@ -0,0 +1,79 @@
+
+const addQueryParams = ( uri: string, params?: URLSearchParams | null ) => {
+	if ( params ) {
+		return uri + '?' + params.toString();
```
What if the URL has query parameters already?

gabrielcaires replied:
I respected the [original behavior](https://github.com/Automattic/wp-calypso/pull/91935/files#diff-1a5cb229a76dfccdc8b471efedc2ab09f2385c9ba00ae3211ba65452ba1c7680L102-L105) as described in these [tests](https://github.com/Automattic/wp-calypso/blob/aef0abfe8e97a0de922621a079f1374562d3ce23/client/landing/stepper/declarative-flow/internals/hooks/use-flow-navigation/test/index.tsx#L84-L101)

The original behavior is: 
- New URL has queryParams, use them and ignore the existing ones.
- New URL doesn't have queryParams, use the existent ones.

alshakero commented on client/landing/stepper/declarative-flow/internals/hooks/use-flow-navigation/index.tsx:
```diff
@@ -0,0 +1,79 @@
+		navigate: customNavigate,
+		params: {
+			flow,
+			step: currentStep,
+			lang,
```
I'd call this `stepSlug` to avoid confusion with the step object.

gabrielcaires replied:
Fixed here 439ffb2d3e865cbe9bd71d38cba5c88146db02a8

alshakero commented on client/landing/stepper/declarative-flow/internals/hooks/use-flow-navigation/index.tsx:
```diff
@@ -0,0 +1,79 @@
+		navigate: customNavigate,
+		params: {
+			flow,
+			step: currentStep,
+			lang,
```
`lang` can probably be deleted?

gabrielcaires replied:
Remove here efc7f1edfbf895951e6976f122b308d3a33772a2

alshakero commented on client/landing/stepper/index.tsx:
```diff
@@ -65,22 +65,31 @@ interface AppWindow extends Window {
+		return ( window.location.href = `/setup/${ DEFAULT_FLOW }${ window.location.search }` );
 	}
 
+	flowName === IMPORT_HOSTED_SITE_FLOW && addHotJarScript();
```
Small favour, can we please make this an `if`? It confused the hell out of me for a good minute.

gabrielcaires replied:
Sure!  2d3b51c0fc6b24ae722561003f79f96335db5918

alshakero reviewed (changes requested):
Beautiful beautiful work. Thank you so much. Left a few comments.

gabrielcaires committed:
Improve hotjar logic

gabrielcaires committed:
Remove unused stepPath

alshakero commented on client/landing/stepper/declarative-flow/internals/components/redirect-to-step/index.tsx:
```diff
@@ -0,0 +1,21 @@
+}
+
+export const RedirectToStep: FC< Props > = ( { slug } ) => {
+	const lang = useFlowLocale();
+	const { flow } = useParams();
+
+	const to = generatePath( '/:flow/:step/:lang?', {
+		flow: flow!,
+		step: slug,
+		lang: lang === 'en' ? null : lang,
+	} );
```
I think we should only add the `lang` param if the user is logged out. It doesn't do anything for logged-in users.

gabrielcaires replied:
I think it makes sense, I implemented it and add some automated tests  [052175e](https://github.com/Automattic/wp-calypso/pull/91935/commits/052175ee334c7ec28e1dc41f9a1a453f04202acc)

gabrielcaires committed:
Improve variable naming

gabrielcaires committed:
Use useCallback to avoid rerenders

gabrielcaires committed:
Remove lang to avoid confusions

alshakero reviewed (approved):
Tested with a Spanish user while logged in. Tested with French and Spanish path names while logged out. Works great.

gabrielcaires committed:
Ignore the language on URL when the user is logged in

gabrielcaires committed:
Fix code style

gabrielcaires committed:
Use const to increase readability
